### PR TITLE
[CR] Two Files API-related fixes

### PR DESCRIPTION
**Pull request by felliott**
The attached patches fix two separate issues with the files API.

1.) d181651 updates NodeFilesList to handle the new JSON-API conformant WaterButler responses.  Relevant: [#OSF-4667]

2.) d65c61d fixes a permissions bug for file checkouts.  It wasn't checking whether or not the request was safe and ergo declining all non-osfstorage files.  Relevant: [#OSF-4674]

The Jira ticket for this task is https://openscience.atlassian.net/browse/OSF-4674

### CODE REVIEW

File: api/files/permissions.py [hunk 1]
```diff
@@ -6,6 +6,9 @@
 class CheckedOutOrAdmin(permissions.BasePermission):
     def has_object_permission(self, request, view, obj):
         assert isinstance(obj, FileNode), 'obj must be a FileNode, got {}'.format(obj)
+        if request.method in permissions.SAFE_METHODS:
+            return True
```
Comment (brianjgeiger) [inline api/files/permissions.py]: Is this the case? Probably is (thinking my way through now). Even if the file is checked out, you should be able to read it, so should be fine. That's what you were thinking?

Comment (felliott) [inline api/files/permissions.py]: Yep. (or more accurately, what @chrisseto was thinking)

File: api/nodes/views.py [hunk 2]
```diff
@@ -67,13 +67,14 @@ class WaterButlerMixin(object):
...
-            item['provider'],
-            FileNode.FOLDER if item['kind'] == 'folder'
+            attrs['provider'],
+            FileNode.FOLDER if attrs['kind'] == 'folder'
             else FileNode.FILE
-        ).get_or_create(self.get_node(check_object_permissions=False), item['path'])
+        ).get_or_create(self.get_node(check_object_permissions=False), attrs['path'])
 
-        file_node.update(None, item, user=self.request.user)
+        file_node.update(None, attrs, user=self.request.user)
```
Comment (brianjgeiger) [inline api/nodes/views.py]: This updates the metadata whenever we get the object info, yes?

File: tests/api_tests/nodes/test_views.py [hunk 3]
```diff
@@ -2787,9 +2787,13 @@ def prepare_mock_wb_response(
     }
 
     if len(files):
-        data = [dict(default_file, **each) for each in files]
+        data = map(
```
Comment (brianjgeiger) [inline tests/api_tests/nodes/test_views.py]: I am having to think far too much about this. What is it supposed to be doing, precisely?

Comment (felliott) [inline tests/api_tests/nodes/test_views.py]: each member of `files` needs to become `{ 'attributes': each }`

**Comment by brianjgeiger** [comment]
By and large, looks good. I expect that the map/lambda/set-comprehension is going to have to be re-written into something more readable. I didn't try to run it, so I'll let future code reviewers do that, but I wanted to at least give it a once-over.

**Comment by felliott** [comment]
Cool, thanks!   I'll update the test and repush.

File: api/files/permissions.py (hunk 1 above)
Comment (chrisseto) [inline api/files/permissions.py]: :+1:

File: tests/api_tests/nodes/test_views.py [hunk 4]
```diff
@@ -2790,11 +2790,16 @@ def prepare_mock_wb_response(
         data = [dict(default_file, **each) for each in files]
     else:
         data = [default_file]
+
+    jsonapi_data = []
+    for datum in data:
```
Comment (brianjgeiger) [inline tests/api_tests/nodes/test_views.py]: :+1: So much better (for my purposes).

**Comment by cwisecarver** [comment]
Sorry it took so long @felliott, got distracted by multithreaded django servers. You know how it is.

It looks fine to me. I double-checked that the real permissions stuff gets called after checkin/checkout just to familiarize myself and make sure we were just letting people in.

I agree with @brianjgeiger about the list comp with the splat being a little brain melting but it looks good now.

Ran the tests. Hit the endpoints. Everything makes sense to me.

**Pull request merged**